Update password endpoint for byron

paweljakubas opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#1483 

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have added `ByronWalletPasspphraseData`
- [x] I have added endpoint to Api modules
- [x] I have added endpoint to swagger
- [x] I have updated Api.TypesSpec and Malformed
- [x] I have added integration tests for updating valid paswd->valid passwd
- [x] I have introduced Password "byron-raw"
- [x] I have enabled null passwords in `ByronWalletPostData` 
- [x] I have updated the rest of codebase bacause of it
- [x] I have added integration tests for updating no passwd -> valid passwd 

# Comments

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Assign the PR to a corresponding milestone
 ✓ Acknowledge any changes required to the Wiki
-->

paweljakubas committed:
add ByronWalletPutPassphraseData

paweljakubas committed:
introduce ApiByronWalletPutPassphraseData

paweljakubas committed:
add PutByronWalletPassphrase and propagate changes throught the codebase

paweljakubas committed:
update Api.Client and CLI because of ByronWalletPassphraseData

paweljakubas committed:
add endpoint to swagger plus update Malformed

paweljakubas committed:
add illustrative integration tests for valid passwd -> valid passwd

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -469,6 +469,11 @@ data WalletPutPassphraseData = WalletPutPassphraseData
     } deriving (Eq, Generic, Show)
 
+data ByronWalletPutPassphraseData = ByronWalletPutPassphraseData
+    { oldPassphrase :: !(Maybe (ApiT (Passphrase "raw")))
```
This won't work unfortunately ... not only does the old passphrase can be empty, but it can also be less than 10 characters (which the new backend doesn't have as a limitation). 

My suggestion would be to have a `ApiT (Passphrase "byron-raw")`, for which we can define a different minimum length that is different from `Passphrase "raw"`, and coerce the types in the API handlers.

paweljakubas replied:
done

KtorZ commented on lib/cli/src/Cardano/CLI.hs:
```diff
@@ -1045,22 +1047,39 @@ cmdWalletUpdatePassphrase mkClient =
+
+instance PassphraseUpdate WalletPutPassphraseData where
+    producePassphraseData (OldNewPassword old new) =
+        WalletPutPassphraseData (ApiT old) (ApiT new)
```
The extra parameter and instances above bring quite some overhead. I'd suggest to leave the command as it is. It won't be possible to update legacy wallets that have no or a invalid passphrase set via the command-line. That's an acceptable limitation :+1:

paweljakubas replied:
reverted support for CLI

paweljakubas committed:
add Password byron-raw

paweljakubas committed:
make Api.TypeSpec passing and revert support for CLI

paweljakubas committed:
add more illustrative tests for empty passphrases

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Client.hs:
```diff
@@ -228,7 +229,7 @@ byronWalletClient =
             , postWallet = _postWallet
             , putWallet = _putWallet
-            , putWalletPassphrase = error "TODO: putWalletPassphrase"
+            , putWalletPassphrase = error "putWalletPassphrase: not impl for byron"
```
This should be `_putByonWalletPassphrase` now, no :) ?

paweljakubas replied:
I have this problem of using `ByronWalletPutPassphraseData` and not `WalletPutPassphraseData` in `WalletClient` and it contributed to the solution you did not liked  done in commit https://github.com/input-output-hk/cardano-wallet/pull/1495/commits/e485e7bad1fd04c6fab3b7997715a354b0d60e85

KtorZ replied:
Okay, I'll look into this :+1:

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -899,7 +908,11 @@ postRandomWallet ctx body = do
+        Nothing ->
+            preparePassphrase EncryptWithPBKDF2 $ Passphrase ""
+        Just (ApiT (Passphrase passwd)) ->
+            preparePassphrase EncryptWithPBKDF2 $ Passphrase passwd
```
suggestion:

```
pwd = preparePassphrase EncryptWithPBKDF2 $ fromMaybe mempty $ getApiT (body ^. #passphrase)
```

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Types.hs:
```diff
@@ -430,7 +430,7 @@ data SomeByronWalletPostData
     { mnemonicSentence :: !(ApiMnemonicT mw)
     , name :: !(ApiT WalletName)
-    , passphrase :: !(ApiT (Passphrase "raw"))
+    , passphrase :: !(Maybe (ApiT (Passphrase "byron-raw")))
```
Why this change? The "standard" API shouldn't allow creating Byron wallets with an empty passphrase, this should only be possible through the legacy endpoint when providing the encrypted XPrv. Sorry for the confusion, I misread your question on slack :(, I thought you were talking about the put passphrase endpoint.

paweljakubas replied:
ok, then we cannot test it easily. I mean now we can and it passes but I can remove the last two commits and should be fine

KtorZ replied:
~~Hmmm. Okay, let's keep it.~~ .. That's a tricky problem. Because now, we need to change the passphrase type on every endpoint that requires a passphrase.

KtorZ commented on lib/core/src/Cardano/Wallet/Api/Server.hs:
```diff
@@ -1136,6 +1161,27 @@ putWalletPassphrase ctx (ApiT wid) body = do
+                W.updateWalletPassphrase wrk wid (old, new)
+            _ -> do
+                let old = Passphrase @"raw" ""
+                W.updateWalletPassphrase wrk wid (old, new)
```
suggestion:

```hs
let old = maybe mempty (coerce . getApiT) oldM  
W.updateWalletPassphrase wrk wid (old, new)
```

KtorZ committed:
implement passphrase update in the byron API client

KtorZ committed:
use XPrv endpoint to create fixture wallet with invalid passphrases

This to avoid having to accept invalid passphrases from the other more recommended way of creating a wallet. For the legacy import, we do not have much choice since the passphrase was already created. But for new wallets (even legacy), we want to enforce a passphrase no matter what.

KtorZ committed:
rewrite update passphrase for byron wallets in a more concise way

KtorZ committed:
re-generate nix machinery
